_posts/2025-05-05-Exposing-WASM-binaries-as-MCP-tools.adoc

```diff
@@ -0,0 +1,125 @@
+It is based on the https://github.com/extism/chicory-sdk[chicory-sdk], an experimental Java SDK using the https://github.com/dylibso/chicory[JVM-native Chicory] runtime.
+
+=== Configuration Methods
+
+You can configure these binaries using either the WildFly CLI or by directly editing the XML configuration (`standalone.xml`).
+
+==== Using CLI
+
+Add a WASM binary definition via the CLI like this:
+
+[source,bash]
+----
+/subsystem=wasm/wasm-tool=my-wasm-tool:add(path="/path/to/your/binary.wasm")
+----
+
+*   `my-wasm-tool`: A unique name for this binary definition.
+*   `path`: The location of the `.wasm` file. This path is relative to the `relative-to` attribute if specified.
+
+==== Using XML Configuration
+
+Alternatively, add the definition within the `<subsystem xmlns="urn:wildfly:wasm:1.0">` element in your configuration file:
+
+[source,xml]
+----
+<subsystem xmlns="urn:jboss:domain:wasm:1.0">
+    <wasm-tool name="my-wasm-tool" path="path/relative/to/module/root/binary.wasm"/>
+    <!-- Other binary definitions -->
+</subsystem>
+----
+
+Ensure the `path` is correctly specified relative to the root of the deployment module.
+
+== Next Steps: CDI Integration
+
+Once defined in the WASM subsystem, these binaries can be injected using CDI.
+The first way you can use them is to get them by name as an `org.wildfly.wasm.api.WasmInvoker` using the `@WasmTool` annotation.
+This very basic interface wraps te call to the binary, passing the parameters as a byte array and getting the result as a byte array.
+
+[source,java]
+----
+
+    @Inject
+    @WasmTool(value = "greet")
+    org.wildfly.wasm.api.WasmInvoker greet;
+
+    public String greet(String person) {
+        byte[] output = greet.call("greet", person.getBytes(StandardCharsets.UTF_8));
+        return new String(output, StandardCharsets.UTF_8);
+    }
+
+----
+
+A second way is to use the `@WasmToolService` over an interface to create a dynamic proxy wrapping the call to the WASM binary and providing a way to serialize and deserialize your arguments to the byte array required by the invoker.
+
+[source,java]
+----
+@WasmToolService(wasmToolConfigurationName = "pizza", wasmMethodName = "retrievePizzeriaAddresses", argumentSerializer = CityJsonSerializer.class)
+public interface Pizza {
+
+    public String pizzas(String city);
+}
+----
+Here you can see that we are calling the exported method `retrievePizzeriaAddresses` on the WASM binary registered with the name *pizza* and the `CityJsonSerializer`class is in charge with converting the java.lang.String city to a JSON object to be consumed by the WASM binary.
+
+Once defined in the WASM subsystem, these binaries can be injected using CDI and subsequently exposed as MCP tools, which we'll cover in the nex step.
+
+
+==  Next Steps: MCP Integration
+
+ Well this is the simplest part as we 'just' have to use our MCP annotations and add the `mcp`subsystem to enable that.
```

Comment:
"This is the simplest part..." [...] "add the `mcp` subsystem (also provided by the WildFly AI Feature Pack)"